### ISSUE TITLE
upsert-resource function can not be used in `fn render`

### ISSUE THREAD

**Issue by droot** [GoogleContainerTools/kpt#3285]
Recent change that makes `functionConfig` available as a regular resource in function input in `fn render` and `fn eval` by default is at odd with the `upsert-resource` function behavior.

The main issue arises because `upsert-resource` uses a resource itself as input that it plans to stamp out so now there is no way to differentiate between the input to the function and the resources in the package and that results in violation of duplicate resource invariant and results in following failure:

```sh
e2etest $ go test -v -run TestE2E/../../examples/upsert-resource-simple
=== RUN   TestE2E
=== RUN   TestE2E/../../examples/upsert-resource-simple
=== PAUSE TestE2E/../../examples/upsert-resource-simple
=== CONT  TestE2E/../../examples/upsert-resource-simple
    runner.go:76: Using kpt binary: /Users/sunilarora/go/bin/kpt
    runner.go:256: Running test against package upsert-resource-simple
    runner.go:338: running command "/Users/sunilarora/go/bin/kpt fn render /var/folders/4r/2mzp5m0d5pvcwgz1r2q2b_g400f698/T/kpt-pipeline-e2e-927888262/upsert-resource-simple --image-pull-policy never"
    runner.go:346: kpt error, stdout: ; stderr: Package "upsert-resource-simple":
        [RUNNING] "gcr.io/kpt-fn/upsert-resource:unstable"
        [FAIL] "gcr.io/kpt-fn/upsert-resource:unstable" in 2.8s
        Error: resource at path "fn-config.yaml" and index "0" already exists
    e2e_test.go:67: failed when running test: actual exit code 1 doesn't match expected 0
--- FAIL: TestE2E (1.84s)
    --- FAIL: TestE2E/../../examples/upsert-resource-simple (3.92s)
FAIL
exit status 1
FAIL    github.com/GoogleContainerTools/kpt-functions-catalog/tests/e2etest     5.981s
e2etest $
```

Haven't thought the solutions yet, but wanted to document this.

@mengqiy we will have to disable the tests for `upsert-resource` (atleast the declarative ones) from `kpt-fn-catalog` for now.

**Comment by mengqiy** [comment]
> we will have to disable the tests for upsert-resource (atleast the declarative ones) from kpt-fn-catalog for now.

Disabling the test can fix the CI. But what makes me worried is that a new kpt release may breaking existing users of `upsert-resource` functions (e.g. the blueprints team).
WDYT?

**Comment by droot** [comment]
> > we will have to disable the tests for upsert-resource (atleast the declarative ones) from kpt-fn-catalog for now.
> 
> Disabling the test can fix the CI. But what makes me worried is that a new kpt release may breaking existing users of `upsert-resource` functions (e.g. the blueprints team). WDYT?

I think the blueprints team uses these function in `out-of-place` mode so this is not a problem where the resource being upserted doesn't exists, so it's not a problem. @bharathkkb did run the release `beta.15` for existing blueprints in their CI.

**Comment by mengqiy** [comment]
It seems the change in https://github.com/GoogleContainerTools/kpt/pull/3228 only break the declarative workflow for `upsert-resource` function.
In that case we can delete the declarative example and update the usage doc for the function saying it only supports imperative workflow.

**Comment by bharathkkb** [comment]
@droot I don't believe we use this function in any of our blueprints, but there maybe other teams using it (PMed info)
https://github.com/GoogleCloudPlatform/blueprints/search?q=upsert-resource

**Comment by droot** [comment]
> It seems the change in #3228 only break the declarative workflow for `upsert-resource` function. In that case we can delete the declarative example and update the usage doc for the function saying it only supports imperative workflow.

I agree, the use case is for generating a resource and that makes sense only for an imperative workflow.

**Comment by andsens** [comment]
I want to use upsert to replace a resource generated from a helm chart, where I need to add some env vars and volumeMounts in a deployment. So I simply duplicate the deployment and change the required fields, which makes the changes easily diffable and updateable when the upstream changes.  
However, I am running into this issue as well, and I'm not quite sure how I am supposed to do this in a declarative workflow without upsert?
